gitlabform/processors/project/job_token_scope_processor.py

```diff
@@ -0,0 +1,213 @@
+from typing import List
+
+from gitlabform.gitlab import GitLab
+from gitlabform.processors import AbstractProcessor
+from cli_ui import warning, info, debug as verbose
+
+from gitlab.v4.objects import Project, ProjectJobTokenScope
+
+
+class JobTokenScopeProcessor(AbstractProcessor):
+    def __init__(self, gitlab: GitLab):
+        super().__init__("job_token_scope", gitlab)
+
+    def _process_configuration(self, project_and_group: str, configuration: dict):
+        job_token_config = configuration.get("job_token_scope", {})
+        verbose(f"Job Token Scope config: {job_token_config}")
+
+        project = self.gl.get_project_by_path_cached(project_and_group)
+        job_token_scope = project.job_token_scope.get()
+
+        self._process_limit_access_to_this_project_setting(
+            job_token_config, job_token_scope
+        )
+
+        allowlist_config = job_token_config.get("allowlist", {})
+        verbose(f"configuration allowlist: {allowlist_config}")
+
+        info("Processing Job Token allowlist")
+
+        enforce = allowlist_config.get("enforce", False)
+
+        self._process_groups(
+            job_token_scope, allowlist_config.get("groups", []), enforce
+        )
+
+        self._process_projects(
+            project, job_token_scope, allowlist_config.get("projects", []), enforce
+        )
+
+    @staticmethod
+    def _process_limit_access_to_this_project_setting(
+        configuration: dict, job_token_scope: ProjectJobTokenScope
+    ):
+        limit_access_to_this_project: bool = configuration.get(
+            "limit_access_to_this_project", True
+        )
+
+        if limit_access_to_this_project != job_token_scope.inbound_enabled:
+            info(
+                f"Updating project job token scope to limit access: {limit_access_to_this_project}"
+            )
+            job_token_scope.enabled = limit_access_to_this_project
+            job_token_scope.save()
+
+            # Refresh object in order for groups_allowlist and allowlist to exist on Scope for modification later
+            job_token_scope.refresh()
```

Comment:
yeah it appears `refresh()` on python-gitlab objects mutates the object itself, on further thought I think I will re-get it outside the function rather than use refresh/mutation, to avoid confusion like this 👍

Reply:
also want to test what happens if I disable token_scope _and_ try remove projects/groups from allowlist 🤔